File: 0493-reverse-pairs/0493-reverse-pairs.java
class Solution {
    public int splitCheck(int[] a, int i, int j) {
        if (i == j) return 0;
        int mid = (i + j) / 2;
        int x = splitCheck(a, i, mid);
        int y = splitCheck(a, mid + 1, j);
        int ans = 0;
        int start1 = i, start2 = mid + 1;
        while (start1 <= mid && start2 <= j) {
            long c = a[start2];
            c += a[start2]; // This seems like an error, it should probably be `c += a[start1];`
            if (a[start1] > c) {
                ans += mid - start1 + 1;
                start2++;
            } else start1++;
        }
        List<Integer> newA = new ArrayList<>();
        start1 = i;
        start2 = mid + 1;
        while (start1 <= mid && start2 <= j) {
            if (a[start1] <= a[start2]) {
                newA.add(a[start1]);
                start1++;
            } else {
                newA.add(a[start2]);
                start2++;
            }
        }
        while (start1 <= mid) newA.add(a[start1++]);
        while (start2 <= j) newA.add(a[start2++]);
        int w = 0;
        for (int z = i; z <= j; z++) {
            a[z] = newA.get(w++);
        }
        return ans + x + y;
    }

    public int reversePairs(int[] nums) {
        return splitCheck(nums, 0, nums.length - 1);
    }
}

File: 0493-reverse-pairs/0493-reverse-pairs.cpp
class Solution {
public:
    int ans = 0;

    void merge(vector<int>& nums, int s, int mid, int e) {
        int cnt = 0;
        for (int i = mid + 1; i <= e; i++) {
            auto it = lower_bound(nums.begin() + s, nums.begin() + mid + 1, 2LL * nums[i] + 1);
            if (it != nums.begin() + mid + 1) {
                int ind = it - nums.begin();
                cnt += (mid - ind + 1);
            }
        }

        ans += cnt;

        vector<int> merger;
        int i = s, j = mid + 1;
        while (i <= mid && j <= e) {
            if (nums[i] < nums[j]) {
                merger.push_back(nums[i++]);
            } else {
                merger.push_back(nums[j++]);
            }
        }
        while (i <= mid) merger.push_back(nums[i++]);
        while (j <= e) merger.push_back(nums[j++]);

        for (int k = 0; k < merger.size(); ++k)
            nums[s + k] = merger[k];
    }

    void dnc(vector<int>& nums, int s, int e) {
        if (s >= e) return;
        int mid = (s + e) / 2;
        dnc(nums, s, mid);
        dnc(nums, mid + 1, e);
        merge(nums, s, mid, e);
    }

    int reversePairs(vector<int>& nums) {
        dnc(nums, 0, nums.size() - 1);
        return ans;
    }
};
